feat: add syscalls using Barreto-Naehrig (BN) curve construction

Flouse opened the pull request:
Introduce a [pairing cryptography](https://en.wikipedia.org/wiki/Pairing-based_cryptography) library to provide [the functions](https://github.com/nervosnetwork/godwoken/blob/1b9de5231b0efdac1e56fd82a2d734dd8bbccf60/crates/generator/src/syscalls/bn.rs) of Barreto-Naehrig (BN) curve such as:

```rust
bn::add
bn::mul
bn::pairing
```

Flouse committed:
feat: add syscalls using Barreto-Naehrig (BN) curve construction

Flouse committed:
config: add default cycles of BN syscalls

jjyr commented on crates/generator/src/syscalls/mod.rs:
```diff
@@ -524,6 +529,76 @@ impl<'a, 'b, S: State, C: ChainView, Mac: SupportMachine> Syscalls<Mac>
+                    if k > 0 {
+                        // Subtract additional cycles per pairing
+                        // https://eips.ethereum.org/EIPS/eip-1108
+                        let additional_cycles = k * 34000 * 3;
```
Can we add this modifier factor to `SyscallCyclesConfig`?

Flouse replied:
https://github.com/godwokenrises/godwoken/pull/808/commits/61b58b1d232570f054345c55c7fcf821338607c8

jjyr commented on crates/config/src/config.rs:
```diff
@@ -560,10 +560,13 @@ pub struct SyscallCyclesConfig {
 
 impl SyscallCyclesConfig {
-    pub fn all_zero() -> Self {
+    pub fn default() -> Self {
```
Please modify the default config according to mainnet settings.

Flouse replied:
https://github.com/godwokenrises/godwoken/pull/808/commits/ead4e4faf3b36f0f5e08df9fa51600f774350e4b

Flouse committed:
chore: update default values of SyscallCyclesConfig according to latest mainnet_v1 settings

> https://github.com/godwokenrises/godwoken-info/blob/bbad7c4/mainnet_v1/gw-mainnet_v1-config-readonly.toml#L138-L150

Flouse committed:
chore: add sys_bn_per_pairing_cycles into SyscallCyclesConfig

jjyr commented:
Please fix the CI

Flouse committed:
test: tune up the max_cycles_limit value in tests

Since we changed the default value of SyscallCyclesConfig, we should
also adjust the limit value in the test.

Flouse commented:
/itest

Flouse commented:
/itest
